external/storm-kafka-client/src/main/java/org/apache/storm/kafka/spout/KafkaSpout.java

```diff
@@ -0,0 +1,458 @@
+    public void deactivate() {
+        shutdown();
+    }
+
+    @Override
+    public void close() {
+        shutdown();
+    }
+
+    private void shutdown() {
+        try {
+            kafkaConsumer.wakeup();
+            if (!consumerAutoCommitMode) {
+                commitOffsetsForAckedTuples();
+            }
+        } finally {
+            //remove resources
+            kafkaConsumer.close();
+        }
+    }
+
+    @Override
+    public void declareOutputFields(OutputFieldsDeclarer declarer) {
+        kafkaSpoutStreams.declareOutputFields(declarer);
+    }
+
+    @Override
+    public String toString() {
+        return "{acked=" + acked + "} ";
+    }
+
+    // ======= Offsets Commit Management ==========
+
+    private static class OffsetComparator implements Comparator<MessageId> {
+        public int compare(MessageId m1, MessageId m2) {
+            return m1.offset() < m2.offset() ? -1 : m1.offset() == m2.offset() ? 0 : 1;
+        }
+    }
+
+    /**
+     * This class is not thread safe
+     */
+    private class OffsetEntry {
+        private final TopicPartition tp;
+        private long committedOffset;               // last offset committed to Kafka, or initial fetching offset (initial value depends on offset strategy. See KafkaSpoutConsumerRebalanceListener)
+        private final NavigableSet<MessageId> ackedMsgs = new TreeSet<>(OFFSET_COMPARATOR);     // acked messages sorted by ascending order of offset
+
+        public OffsetEntry(TopicPartition tp, long committedOffset) {
+            this.tp = tp;
+            this.committedOffset = committedOffset;
+            LOG.debug("Created OffsetEntry for [topic-partition={}, committed-or-initial-fetch-offset={}]", tp, committedOffset);
+        }
+
+        public void add(MessageId msgId) {          // O(Log N)
+            ackedMsgs.add(msgId);
+        }
+
+        /**
+         * @return the next OffsetAndMetadata to commit, or null if no offset is ready to commit.
+         */
+        public OffsetAndMetadata findNextCommitOffset() {
+            boolean found = false;
+            long currOffset;
+            long nextCommitOffset = committedOffset;
+            MessageId nextCommitMsg = null;     // this is a convenience variable to make it faster to create OffsetAndMetadata
+
+            for (MessageId currAckedMsg : ackedMsgs) {  // complexity is that of a linear scan on a TreeMap
+                if ((currOffset = currAckedMsg.offset()) == 0 || currOffset != nextCommitOffset) {      // The first poll includes the last committed offset. This if avoids duplication
+                    if (currOffset == nextCommitOffset || currOffset == nextCommitOffset + 1) {            // found the next offset to commit
+                        found = true;
+                        nextCommitMsg = currAckedMsg;
+                        nextCommitOffset = currOffset;
+                        LOG.trace("Found offset to commit [{}]. {}", currOffset, toString());
```

Comment:
can you pass `this` instead of toString() so that toString() is called only if trace is enabled.